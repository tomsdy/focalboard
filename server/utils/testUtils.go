package utils

import "github.com/stretchr/testify/mock"

var Anything = mock.MatchedBy(func(interface{}) bool { return true })
