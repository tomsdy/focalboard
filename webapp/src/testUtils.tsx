// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import {IntlProvider} from 'react-intl'
import React from 'react'

export const wrapIntl = (children: any) => <IntlProvider locale='en'>{children}</IntlProvider>

export function mockDOM(): void {
    window.focus = jest.fn()
    document.createRange = () => {
        const range = new Range()

        range.getBoundingClientRect = jest.fn()

        range.getClientRects = () => {
            return {
                item: () => null,
                length: 0,
                [Symbol.iterator]: jest.fn(),
            }
        }

        return range
    }
}
