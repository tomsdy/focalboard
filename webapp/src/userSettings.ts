// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

export class UserSettings {
    static get prefillRandomIcons(): boolean {
        return localStorage.getItem('randomIcons') !== 'false'
    }

    static set prefillRandomIcons(newValue: boolean) {
        localStorage.setItem('randomIcons', JSON.stringify(newValue))
    }

    static get preferredDateFormat(): string {
        return localStorage.getItem('preferredDateFormat') || 'MM/DD/YYYY'
    }

    static set preferredDateFormat(newValue: string) {
        localStorage.setItem('preferredDateFormat', newValue)
    }
}

const keys = ['language', 'theme', 'lastBoardId', 'lastViewId', 'emoji-mart.last', 'emoji-mart.frequently', 'randomIcons', 'preferredDateFormat']

export function exportUserSettingsBlob(): string {
    return window.btoa(exportUserSettings())
}

function exportUserSettings(): string {
    const settings = Object.fromEntries(keys.map((key) => [key, localStorage.getItem(key)]))
    settings.timestamp = `${Date.now()}`
    return JSON.stringify(settings)
}

export function importUserSettingsBlob(blob: string): boolean {
    return importUserSettings(window.atob(blob))
}

function importUserSettings(json: string): boolean {
    const settings = parseUserSettings(json)
    const timestamp = settings.timestamp
    const lastTimestamp = localStorage.getItem('timestamp')
    if (!timestamp || (lastTimestamp && Number(timestamp) <= Number(lastTimestamp))) {
        return false
    }
    for (const [key, value] of Object.entries(settings)) {
        if (value) {
            localStorage.setItem(key, value as string)
        } else {
            localStorage.removeItem(key)
        }
    }
    return true
}

function parseUserSettings(json: string): any {
    try {
        return JSON.parse(json)
    } catch (e) {
        return {}
    }
}
