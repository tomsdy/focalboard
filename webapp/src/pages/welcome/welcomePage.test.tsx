// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import React from 'react'

import {render, screen} from '@testing-library/react'

import {createMemoryHistory} from 'history'

import {Router} from 'react-router-dom'

import userEvent from '@testing-library/user-event'

import {UserSettings} from '../../userSettings'

import {wrapIntl} from '../../testUtils'

import WelcomePage from './welcomePage'

beforeEach(() => {
    UserSettings.welcomePageViewed = null
})

describe('pages/welcome', () => {
    const history = createMemoryHistory()
    test('Welcome Page shows Explore Page', () => {
        const {container} = render(wrapIntl(
            <Router history={history}>
                <WelcomePage/>
            </Router>,
        ))
        expect(screen.getByText('Explore')).toBeDefined()
        expect(container).toMatchSnapshot()
    })

    test('Welcome Page shows Explore Page And Then Proceeds after Clicking Explore', () => {
        history.replace = jest.fn()
        render(wrapIntl(
            <Router history={history}>
                <WelcomePage/>
            </Router>,
        ))
        const exploreButton = screen.getByText('Explore')
        expect(exploreButton).toBeDefined()
        userEvent.click(exploreButton)
        expect(history.replace).toBeCalledWith('/dashboard')
    })

    test('Welcome Page does not render explore page the second time we visit it', () => {
        history.replace = jest.fn()
        UserSettings.welcomePageViewed = 'true'
        render(wrapIntl(
            <Router history={history}>
                <WelcomePage/>
            </Router>,
        ))
        expect(history.replace).toBeCalledWith('/dashboard')
    })

    test('Welcome Page redirects us when we have a r query parameter with welcomePageViewed set to true', () => {
        history.replace = jest.fn()
        history.location.search = 'r=123'
        UserSettings.welcomePageViewed = 'true'
        render(wrapIntl(
            <Router history={history}>
                <WelcomePage/>
            </Router>,
        ))
        expect(history.replace).toBeCalledWith('123')
    })

    test('Welcome Page redirects us when we have a r query parameter with welcomePageViewed set to null', () => {
        history.replace = jest.fn()
        history.location.search = 'r=123'
        render(wrapIntl(
            <Router history={history}>
                <WelcomePage/>
            </Router>,
        ))
        const exploreButton = screen.getByText('Explore')
        expect(exploreButton).toBeDefined()
        userEvent.click(exploreButton)
        expect(history.replace).toBeCalledWith('123')
    })
})
