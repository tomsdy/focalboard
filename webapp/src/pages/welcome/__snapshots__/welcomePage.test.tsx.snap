// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages/welcome Welcome Page shows Explore Page 1`] = `
<div>
  <div
    class="WelcomePage"
  >
    <div>
      <h1
        class="text-heading9"
      >
        Welcome To Boards
      </h1>
      <div
        class="WelcomePage__subtitle"
      >
        Boards is a project management tool that helps define, organize, track and manage work across teams, using a familiar kanban board view
      </div>
      <img
        alt="Boards Welcome Image"
        class="WelcomePage__image WelcomePage__image--large"
        src="test-file-stub"
      />
      <img
        alt="Boards Welcome Image"
        class="WelcomePage__image WelcomePage__image--small"
        src="test-file-stub"
      />
      <button
        class="Button filled size--large"
        type="button"
      >
        <span>
          Explore
        </span>
        <i
          class="CompassIcon icon-chevron-right Icon Icon--right"
        />
      </button>
    </div>
  </div>
</div>
`;
